From fdd22658bcd63cc4d6c644f53f164e802436eee4 Mon Sep 17 00:00:00 2001
From: Manish Dharanenthiran <quic_mdharane@quicinc.com>
Date: Mon, 27 May 2024 15:13:15 +0530
Subject: [PATCH 6/6] wifi: mac80211: Add WMM action frame in broadcast frame
 allowed list

In ieee80211_mgmt_tx, for certain action frame category types 'sta'
object is not mandatory in-order to allow broadcast or multicast frames.

Add 'WLAN_CATEGORY_WNM' action category to the list as BSS Transition
management query frame will be transmitted as a broadcast frame when
ML reconfigure link removal is scheduled.

Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 net/mac80211/offchannel.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

--- a/net/mac80211/offchannel.c
+++ b/net/mac80211/offchannel.c
@@ -862,7 +862,9 @@ int ieee80211_mgmt_tx(struct wiphy *wiph
 		if (!ieee80211_is_action(mgmt->frame_control) ||
 		    mgmt->u.action.category == WLAN_CATEGORY_PUBLIC ||
 		    mgmt->u.action.category == WLAN_CATEGORY_SELF_PROTECTED ||
-		    mgmt->u.action.category == WLAN_CATEGORY_SPECTRUM_MGMT) {
+		    mgmt->u.action.category == WLAN_CATEGORY_SPECTRUM_MGMT ||
+		    (is_broadcast_ether_addr(mgmt->da) &&
+		     mgmt->u.action.category == WLAN_CATEGORY_WNM)) {
 			rcu_read_unlock();
 			break;
 		}
